data/data-pipeline/data_pipeline/etl/runner.py

```diff
@@ -32,36 +33,62 @@ def get_datasets_to_run(dataset_to_run: str):
     return dataset_list
 
 
-def etl_runner(dataset_to_run: str = None) -> None:
-    """Runs all etl processes or a specific one
+def build_instance_class_from_dataset(dataset: str = None) -> None:
+    """Runs a specific dataset
 
     Args:
-        dataset_to_run (str): Run a specific ETL process. If missing, runs all processes (optional)
+        dataset (str): A specific dataset eligible for a specific ETL process.
 
     Returns:
         None
     """
-    dataset_list = get_datasets_to_run(dataset_to_run)
+    etl_module = importlib.import_module(
+        f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
+    )
+    etl_class = getattr(etl_module, dataset["class_name"])
+    etl_instance = etl_class()
 
-    # Run the ETLs for the dataset_list
-    for dataset in dataset_list:
-        etl_module = importlib.import_module(
-            f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
-        )
-        etl_class = getattr(etl_module, dataset["class_name"])
-        etl_instance = etl_class()
+    etl_module = importlib.import_module(
+        f"data_pipeline.etl.sources.{dataset['module_dir']}.etl"
+    )
+    etl_class = getattr(etl_module, dataset["class_name"])
+    etl_instance = etl_class()
+
+    # run extract
+    etl_instance.extract()
+
+    # run transform
+    etl_instance.transform()
+
+    # run load
+    etl_instance.load()
+
+    # cleanup
+    etl_instance.cleanup()
 
-        # run extract
-        etl_instance.extract()
 
-        # run transform
-        etl_instance.transform()
+def etl_runner(dataset_to_run: str = None) -> None:
+    """Runs all etl processes or a specific one
+
+    Args:
+        dataset_to_run (str): Run a specific ETL process. If missing, runs all processes (optional)
 
-        # run load
-        etl_instance.load()
+    Returns:
+        None
+    """
+    dataset_list = get_datasets_to_run(dataset_to_run)
 
-        # cleanup
-        etl_instance.cleanup()
+    # Create a pool of processes. By default, one is created for each CPU in your machine.
+    with concurrent.futures.ProcessPoolExecutor() as executor:
+        # Process the list of ETL datasets, but split the work across the
+        # process pool to use all CPUs
+        for dataset, _ in zip(
+            dataset_list,
+            executor.map(
+                build_instance_class_from_dataset, dataset_list
+            ),
```

Comment:
Another idea:

```
from queue import Queue
from threading import Thread

class ETLWorker(Thread):

    def __init__(self, queue):
        Thread.__init__(self)
        self.queue = queue

    def run(self):
        while True:
            # Get the work from the queue and expand the tuple
            dataset = self.queue.get()
            try:
                build_instance_class_from_dataset(dataset)
            finally:
                self.queue.task_done()

    # Create a queue to communicate with the worker threads
    queue = Queue()
    # Create 8 worker threads
    for x in range(8):
        worker = ETLWorker(queue)
        # Setting daemon to True will let the main thread exit even though the workers are blocking
        worker.daemon = True
        worker.start()
    # Put the tasks into the queue as a tuple
    for dataset in datasets:
        logger.info('Queueing {}'.format(dataset))
        queue.put((dataset))
    # Causes the main thread to wait for the queue to finish processing all the tasks
    queue.join()
    logging.info('Took %s', time() - ts)
```

Reply:
@sverchdotgov @lucasmbrown-usds 

Something I should have noted in this PR and in standup:

 When using `map` from `concurrent.futures`, each element from the iterable is submitted separately to the executor, which creates a Future object for each call. It then returns an iterator which yields the results returned by the futures.

Future objects are rather heavyweight, they do a lot of work to allow all the features they provide (like callbacks, ability to cancel, check status, ...).

Compared to that, `multiprocessing.Pool` has much less overhead. It submits jobs in batches (reducing IPC overhead), and directly uses the result returned by the function. For big batches of jobs, multiprocessing is definitely the better options. I thought that since we are not really putting in too many jobs, we could get away with futures.

Futures are wonderful if you want to submit long-running jobs where the overhead isn't that important, where you want to be notified by callback or check from time to time to see if they're done or be able to cancel the execution individually.

Reply:
@sverchdotgov  - I have refactored the `map` and erred on the side of caution.

Reply:
@sverchdotgov I also did not really answer your question. At the time I wrote a [decorator](https://github.com/usds/justice40-tool/pull/1039#issuecomment-992682249) and just through it to profile an individual process. I think I can add more OS level information and make this a utility if @lucasmbrown-usds thinks this would help with future performance optimizations.